statistics/cmsketch.go

```diff
@@ -44,29 +57,227 @@ func NewCMSketch(d, w int32) *CMSketch {
 	return &CMSketch{depth: d, width: w, table: tbl}
 }
 
+// NewCMSketchWithTopN returns a new CM sketch with TopN elements.
+// Total is the size of the whole dataset
+// Only when the number of topn elements exceeded len(data) / topnThreshold will store topn elements
+func NewCMSketchWithTopN(d, w int32, data [][]byte, n uint32, total uint64) *CMSketch {
+	tbl := make([][]uint32, d)
+	for i := range tbl {
+		tbl[i] = make([]uint32, w)
+	}
+	c := &CMSketch{depth: d, width: w, table: tbl}
+	c.BuildTopN(data, n, topNThreshold, total)
+	return c
+}
+
+// BuildTopN builds table of top N elements.
+// elements in data should not be modified after this call.
+// Not exactly n elements, will add a few elements, the number of which is close to the n-th most element.
+func (c *CMSketch) BuildTopN(data [][]byte, n, topnThreshold uint32, total uint64) {
+	sampleSize := uint64(len(data))
+	ratio := total / uint64(sampleSize)
+	if total < sampleSize {
+		ratio = 1
+	}
+
+	counter := make(map[uint64][]cmscount)
+	ndv := uint32(0)
+
+	for k := range data {
+		h1, h2 := murmur3.Sum128(data[k])
+		sl, ok := counter[h1]
+		if !ok {
+			sl = make([]cmscount, 0)
+		}
+		exists := false
+		for i := range sl {
+			if sl[i].h2 == h2 && bytes.Equal(sl[i].data, data[k]) {
+				exists = true
+				sl[i].count++
+			}
+		}
+		if !exists {
+			sl = append(sl, cmscount{h1, h2, data[k], 1})
+			ndv++
+		}
+		counter[h1] = sl
+	}
+
+	sorted := make([]uint64, ndv)
+	seq := 0
+
+	for _, sl := range counter {
+		for i := range sl {
+			sorted[seq] = sl[i].count
+			seq++
+		}
+	}
+
+	sort.Slice(sorted, func(i, j int) bool {
+		return sorted[i] > sorted[j]
+	})
+
+	if n > ndv {
+		n = ndv
+	}
+
+	topnCountThreshold := sorted[n-1]
+	newTopN := topnCountThreshold
+	realTopN := n
+
+	for i := n; i < ndv && i < n*2; i++ {
+		if sorted[i]*2 < topnCountThreshold {
+			break
+		}
+		newTopN = sorted[i]
+		realTopN = i
+	}
+
+	topnCountThreshold = newTopN
+
+	topn := make([]cmscount, realTopN)
+
+	topncount := uint64(0)
+
+	seq = 0
+	for _, sl := range counter {
+		for i := range sl {
+			if uint32(seq) < realTopN && sl[i].count >= topnCountThreshold {
+				topn[seq] = sl[i]
+				topn[seq].count *= ratio
```

Comment:
If every value just occurred once, then the `ratio` should be 1, not `total / uint64(sampleSize)`. It should take the ndv of the whole column into consideration.